[Text Matching] Improve performance of Editor::countMatchesForText

yulunwu2 opened the pull request:
#### f64b87c276020871c1c34becc730b6d61b068a0f
<pre>
[Text Matching] Improve performance of Editor::countMatchesForText
<a href="https://bugs.webkit.org/show_bug.cgi?id=299571">https://bugs.webkit.org/show_bug.cgi?id=299571</a>
&lt;<a href="https://rdar.apple.com/160011859">rdar://160011859</a>&gt;

Reviewed by Ryosuke Niwa.

This PR improves the performance of countMatchesForText() by replacing
looped calls to findPlainText() to findAllPlainText() which finds all
matches in a single pass using forEachMatch.

This PR also inlines construction of SimpleRange instead of relying on
the expensive rangeForMatch() function used by findPlainText.

Finally, this PR allows us to countMatchesForText() while only constructing
a single iterator over the range, where as we were previously making very
expensive calls to rangeForMatch() and constructing/tearing down O(matches)
iterators to build our ranges.

<a href="https://searchfox.org/wubkat/source/Source/WebCore/editing/TextIterator.cpp#2662-2684">https://searchfox.org/wubkat/source/Source/WebCore/editing/TextIterator.cpp#2662-2684</a>

* Source/WebCore/editing/Editor.cpp:
(WebCore::Editor::countMatchesForText):
* Source/WebCore/page/Page.cpp:
(WebCore::Page::findTextMatches):
* Source/WebKit/WebProcess/WebPage/FindController.cpp:
(WebKit::FindController::countStringMatches):
(WebKit::FindController::findStringMatches):

Canonical link: <a href="https://commits.webkit.org/300894@main">https://commits.webkit.org/300894@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit-security/commit/ead53cfa471612edde9d8a07a72f6e684d3454f8

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [⏳ 🧪 style ](https://ews-build.webkit.org/#/builders/Style-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 ios ](https://ews-build.webkit.org/#/builders/iOS-18-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 mac ](https://ews-build.webkit.org/#/builders/macOS-Sonoma-Release-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 wpe ](https://ews-build.webkit.org/#/builders/WPE-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 win ](https://ews-build.webkit.org/#/builders/Win-Build-EWS "Waiting in queue, processing has not started yet") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/4b4131f9-8e4c-4a66-8403-6bbe60a0a249) 
| [⏳ 🧪 bindings ](https://ews-build.webkit.org/#/builders/Bindings-Tests-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 ios-sim ](https://ews-build.webkit.org/#/builders/iOS-18-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 mac-AS-debug ](https://ews-build.webkit.org/#/builders/macOS-Sequoia-Debug-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 wpe-wk2 ](https://ews-build.webkit.org/#/builders/WPE-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 win-tests ](https://ews-build.webkit.org/#/builders/Win-Build-EWS "Waiting in queue, processing has not started yet") | ⏳ 🛠 mac-apple 
| [⏳ 🧪 webkitperl ](https://ews-build.webkit.org/#/builders/WebKitPerl-Tests-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 ios-wk2 ](https://ews-build.webkit.org/#/builders/iOS-18-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 api-mac ](https://ews-build.webkit.org/#/builders/macOS-Sonoma-Release-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 api-wpe ](https://ews-build.webkit.org/#/builders/WPE-Build-EWS "Waiting in queue, processing has not started yet") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/420dbd87-71e6-4282-bd21-f6ca12ba89ba) 
| [⏳ 🧪 webkitpy ](https://ews-build.webkit.org/#/builders/WebKitPy-Tests-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 ios-wk2-wpt ](https://ews-build.webkit.org/#/builders/iOS-18-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 mac-wk1 ](https://ews-build.webkit.org/#/builders/macOS-Sonoma-Release-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 wpe-cairo ](https://ews-build.webkit.org/#/builders/WPE-Cairo-Build-EWS "Waiting in queue, processing has not started yet") | | 
| [⏳ 🛠 🧪 jsc ](https://ews-build.webkit.org/#/builders/JSC-Tests-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 api-ios ](https://ews-build.webkit.org/#/builders/iOS-18-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 mac-wk2 ](https://ews-build.webkit.org/#/builders/macOS-Sonoma-Release-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 gtk ](https://ews-build.webkit.org/#/builders/GTK-Build-EWS "Waiting in queue, processing has not started yet") | | 
| [⏳ 🛠 🧪 jsc-arm64 ](https://ews-build.webkit.org/#/builders/JSC-Tests-arm64-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 vision ](https://ews-build.webkit.org/#/builders/visionOS-2-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 mac-AS-debug-wk2 ](https://ews-build.webkit.org/#/builders/macOS-Sequoia-Debug-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 gtk-wk2 ](https://ews-build.webkit.org/#/builders/GTK-Build-EWS "Waiting in queue, processing has not started yet") | | 
| [⏳ 🧪 services ](https://ews-build.webkit.org/#/builders/Services-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 vision-sim ](https://ews-build.webkit.org/#/builders/visionOS-2-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 mac-wk2-stress ](https://ews-build.webkit.org/#/builders/macOS-Sonoma-Release-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 api-gtk ](https://ews-build.webkit.org/#/builders/GTK-Build-EWS "Waiting in queue, processing has not started yet") | | 
| [  ~~🛠 🧪 merge~~](https://ews-build.webkit.org/#/builders/19/builds/26151 "The change is no longer eligible for processing.") | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-2-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 mac-intel-wk2 ](https://ews-build.webkit.org/#/builders/macOS-Sonoma-Release-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 playstation ](https://ews-build.webkit.org/#/builders/PlayStation-Build-EWS "Waiting in queue, processing has not started yet") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/19515 "Built successfully and passed tests") | [⏳ 🛠 tv ](https://ews-build.webkit.org/#/builders/tvOS-18-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 mac-safer-cpp ](https://ews-build.webkit.org/#/builders/macOS-Safer-CPP-Checks-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 jsc-armv7 ](https://ews-build.webkit.org/#/builders/JSC-ARMv7-32bits-Build-EWS "Waiting in queue, processing has not started yet") | | 
| | [⏳ 🛠 tv-sim ](https://ews-build.webkit.org/#/builders/tvOS-18-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | | [⏳ 🧪 jsc-armv7-tests ](https://ews-build.webkit.org/#/builders/JSC-ARMv7-32bits-Build-EWS "Waiting in queue, processing has not started yet") | | 
| | [⏳ 🛠 watch ](https://ews-build.webkit.org/#/builders/watchOS-11-Build-EWS "Waiting in queue, processing has not started yet") | | | | 
| | [⏳ 🛠 watch-sim ](https://ews-build.webkit.org/#/builders/watchOS-11-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | | | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Source/WebCore/editing/Editor.cpp:
```diff
@@ -4086,42 +4086,18 @@ unsigned Editor::countMatchesForText(const String& target, const std::optional<S
-            if (auto shadowTreeRoot = searchRange->start.container->containingShadowRoot())
-                searchRange->end = makeBoundaryPointAfterNodeContents(*shadowTreeRoot);
+    if (markMatches) {
+        for (const auto& match : allMatches) {
```
No `{` ~ `}` around a single line statement.

yulunwu2 replied:
Done and thank you for the review!

rniwa commented on Source/WebCore/editing/TextIterator.cpp:
```diff
@@ -2705,6 +2705,50 @@ SimpleRange findClosestPlainText(const SimpleRange& range, const String& target,
+            return std::nullopt;
+        auto start = it.range().start;
+        if (match.length > 1) {
+            it.advance(match.length - 1);
```
Where did `-1` come from?

yulunwu2 replied:
Fixed, thank you for the review!

rniwa replied:
Hm... `-1` is still there. Might be good to explain why we're subtracting 1 here.

yulunwu2 replied:
Ah sorry, I was replying to the review as I was editing the code locally. I'll be sure to reply to the comments after uploading in the future. Sorry for the confusion!

rniwa commented on Source/WebCore/editing/TextIterator.cpp:
```diff
@@ -2705,6 +2705,50 @@ SimpleRange findClosestPlainText(const SimpleRange& range, const String& target,
+        matches.append(WTFMove(*foundRange));
+
+        // Check limit after adding match
+        if (limit > 0 && matches.size() >= limit)
```
Why not just `return limit > 0 && matches.size() >= limit`?

yulunwu2 replied:
Done and thank you for the review!

yulunwu2 committed:
[Text Matching] Improve performance of Editor::countMatchesForText
https://bugs.webkit.org/show_bug.cgi?id=299571
<rdar://160011859>

Reviewed by Ryosuke Niwa.

This PR improves the performance of countMatchesForText() by replacing
looped calls to findPlainText() to findAllPlainText() which finds all
matches in a single pass using forEachMatch.

This PR also inlines construction of SimpleRange instead of relying on
the expensive rangeForMatch() function used by findPlainText.

Finally, this PR allows us to countMatchesForText() while only constructing
a single iterator over the range, where as we were previously making very
expensive calls to rangeForMatch() and constructing/tearing down O(matches)
iterators to build our ranges.

https://searchfox.org/wubkat/source/Source/WebCore/editing/TextIterator.cpp#2662-2684

* Source/WebCore/editing/Editor.cpp:
(WebCore::Editor::countMatchesForText):
* Source/WebCore/page/Page.cpp:
(WebCore::Page::findTextMatches):
* Source/WebKit/WebProcess/WebPage/FindController.cpp:
(WebKit::FindController::countStringMatches):
(WebKit::FindController::findStringMatches):

Canonical link: https://commits.webkit.org/300894@main
